### PR TITLE
Cleanup to syntax::print/rustc::hir::print

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This is in preparation for a removal of `io::Result<()>` from all of the printing infrastructure methods (separate PR due to huge diff), along with potentially some other cleanups.

r? @Centril (but feel free to reassign if there's someone better -- there's no functional changes here)

### CODE REVIEW

**Comment by Centril** [comment]
I don't understand the larger picture of excising `io::Result<()>` from pretty printing.
However, assuming you do and the `&mut Vec<u8>` are going towards a goal you understand...
...the changes in the PR look fine.

So r=me on that.

**Comment by Centril** [comment]
That said, cc @petrochenkov and @eddyb.

**Comment by Mark-Simulacrum** [comment]
The larger goal is mostly just a simplification of the code, which may lead to both compile time wins (`?` on every other line of code isn't cheap due to expansion and later processing, I believe) and potentially performance wins. We can hold off on landing this PR until we get numbers, but I think it's pretty much an improvement without any significant downsides.

File: src/libsyntax/print/pp.rs [hunk 1]
```diff
@@ -259,7 +259,7 @@ pub fn mk_printer<'a>(out: Box<dyn io::Write+'a>, linewidth: usize) -> Printer<'
 }
 
 pub struct Printer<'a> {
-    out: Box<dyn io::Write+'a>,
+    out: &'a mut Vec<u8>,
```
Comment (eddyb) [inline src/libsyntax/print/pp.rs]: Why not a `String`?

**Pull request closed without merging**